google/cloud/spanner/internal/connection_impl.cc

```diff
@@ -436,33 +440,47 @@ StatusOr<std::vector<ReadPartition>> ConnectionImpl::PartitionReadImpl(
   *request.mutable_partition_options() = internal::ToProto(partition_options);
 
   auto stub = session_pool_->GetStub(*session);
-  auto response = internal::RetryLoop(
-      retry_policy_prototype_->clone(), backoff_policy_prototype_->clone(),
-      true,
-      [&stub](grpc::ClientContext& context,
-              spanner_proto::PartitionReadRequest const& request) {
-        return stub->PartitionRead(context, request);
-      },
-      request, __func__);
-  if (!response.ok()) {
-    auto status = std::move(response).status();
-    if (internal::IsSessionNotFound(status)) session->set_bad();
-    return status;
-  }
+  for (;;) {
+    auto response = internal::RetryLoop(
+        retry_policy_prototype_->clone(), backoff_policy_prototype_->clone(),
+        true,
+        [&stub](grpc::ClientContext& context,
+                spanner_proto::PartitionReadRequest const& request) {
+          return stub->PartitionRead(context, request);
+        },
+        request, __func__);
+    if (s->has_begin()) {
+      if (response.ok() && response->has_transaction()) {
+        s->set_id(response->transaction().id());
+      } else {
+        // A transaction should have been returned but was not.
+        auto begin = BeginTransaction(session, s->begin(), __func__);
+        if (!begin.ok()) {
+          s = begin.status();  // invalidate the transaction
+          return begin.status();
```

Comment:
And, given that the response was not OK, I think this should return `response.status()`.